src/asio_service.cxx

```diff
@@ -835,11 +841,12 @@ class asio_rpc_listener
                   err.value(), err.message().c_str() );
         }
 
+        std::lock_guard<std::mutex> guard(listener_lock_);
         if (!stopped_) {
             // Re-listen only when not stopped,
             // otherwise crash happens as this class or `acceptor_`
             // may be destroyed in the meantime.
-            this->start();
+            this->start(guard);
```

Comment:
`start` function calls `handle_accept` asynchronously by registering it to `async_accept`. But I'm not 100% sure if such registered `handle_accept` is always executed by different threads or sometimes it can be executed within a single thread immediately. In the latter case, deadlock can happen. So to be safe, just in case, can you please define `listener_lock_` as `std::recursive_mutex`?

Reply:
Good question:
https://www.boost.org/doc/libs/1_76_0/doc/html/boost_asio/reference/basic_socket_acceptor/async_accept/overload1.html
`The function call always returns immediately.`

Reply:
But I can make it `std::recursive_mutex`

Reply:
Ok, then `std::mutex` should be fine.